scenario/smartsearch.go

```diff
@@ -0,0 +1,350 @@
+			},
+		)
+		logSearchResult(sessionState.LogEntry, id, searchResult)
+		return err
+	}, actionState, true, "SearchResults call failed")
+}
+
+func newSearchTextChunks(randomizer helpers.Randomizer, searchText string, pasteSearchText bool) (searchTextChunks, error) {
+	if pasteSearchText {
+		return searchTextChunks{
+			searchTextChunk{
+				Text:                    searchText,
+				PostTypingThinkDuration: 0,
+				TypingDuration:          0,
+			},
+		}, nil
+	}
+
+	const typeOneCharDuration = 300 * time.Millisecond
+	const minPostTypingDuration = 700 * time.Millisecond
+	const maxPostTypingDuration = 1300 * time.Millisecond
+
+	chunks := searchTextChunks{}
+	currentStart := 0
+	for currentEnd := 1; currentEnd < len(searchText); currentEnd++ {
+		randInt, err := randomizer.RandWeightedInt([]int{1, 9})
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to randomize search text chunk")
+
+		}
+		shallSplit := randInt == 0
+		postTypingDuration, err := randomizer.RandDuration(minPostTypingDuration, maxPostTypingDuration)
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to randomize search text chunk post typing duration")
+		}
+		if shallSplit {
+			textChunk := searchText[currentStart:currentEnd]
+			currentStart = currentEnd
+			chunks = append(chunks, searchTextChunk{
+				Text:                    textChunk,
+				PostTypingThinkDuration: postTypingDuration,
+				TypingDuration:          time.Duration(len(textChunk)) * typeOneCharDuration,
+			})
+		}
+	}
+	textChunk := searchText[currentStart:]
+	chunks = append(chunks, searchTextChunk{
+		Text:                    textChunk,
+		PostTypingThinkDuration: 0,
+		TypingDuration:          time.Duration(len(textChunk)) * typeOneCharDuration,
+	})
+	return chunks, nil
+
+}
+
+func (chunks searchTextChunks) simulate() <-chan string {
+	textChan := make(chan string, len(chunks))
+	go func() {
+		currentText := ""
+		for _, chunk := range chunks {
+			currentText = currentText + chunk.Text
+			time.Sleep(chunk.TypingDuration)
+			textChan <- currentText
+			time.Sleep(chunk.PostTypingThinkDuration)
+		}
+		close(textChan)
+	}()
+	return textChan
+}
+
+// Execute implements ActionSettings interface
+func (settings SmartSearchSettings) Execute(sessionState *session.State, actionState *action.State, connection *connection.ConnectionSettings, label string, reset func()) {
+	uplink := sessionState.Connection.Sense()
+	if uplink.CurrentApp == nil {
+		actionState.AddErrors(errors.New("not connected to app"))
+		return
+	}
+	doc := uplink.CurrentApp.Doc
+	rand := sessionState.Randomizer()
+	searchText := settings.SearchTextList[rand.Rand(len(settings.SearchTextList))]
+
+	searchTextChunks, err := newSearchTextChunks(rand, searchText, settings.PasteSearchText)
+	if err != nil {
+		actionState.AddErrors(err)
+		return
+	}
+
+	sessionState.LogEntry.LogDebugf("search text chunks: %+v", searchTextChunks)
+
+	reset()
+
+	cnt := 0
+	for searchText := range searchTextChunks.simulate() {
+		cnt++
```

Comment:
add check for context canceled not to continue looping on execution stopped. Also with added possible error on actionstate as per above comment, a break check on actionState.Failed.